src/sentry/api/endpoints/project_release_files.py

```diff
@@ -152,45 +154,34 @@ def post_releasefile(request, release, logger):
         return Response(serialize(releasefile, request.user), status=201)
 
 
-class ListQuerySet:
-    """Pseudo queryset offering a subset of QuerySet operations"""
+class ArtifactSource:
+    """Provides artifact data to ChainPaginator on-demand"""
 
-    def __init__(self, release_files, query=None):
-        self._files = [
-            # Mimic "or" operation applied for real querysets:
-            rf
-            for rf in release_files
-            if not query or any(search_string.lower() in rf.name.lower() for search_string in query)
-        ]
+    def __init__(self, dist: Optional[Distribution], files: dict, query: List[str]):
+        self._dist = dist
+        self._files = files.items()
+        self._query = query
+        self._ready = False
 
-    def get(self):
-        files = self._files
-        if not files:
-            raise ObjectDoesNotExist
-        if len(files) > 1:
-            raise MultipleObjectsReturned
+    def __len__(self):
+        return len(self._files)
 
-        return files[0]
+    def __getitem__(self, range):
+        self._ensure_ready()
 
-    def annotate(self, **kwargs):
-        if kwargs:
-            raise NotImplementedError
+        return [pseudo_releasefile(url, info, self._dist) for url, info in self._files[range]]
```

Comment:
would do `_ensure_ready` straight in `__init__` or make the sorted list a cached_property, this is hard to follow. for a second I thought you'd index into `dict.items()`

Reply:
Done.